File: advent2017/extras/cxx/src/days/day05.cpp
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "e9c74e01657b99ad1be3cedce52f75bb0e2ac9dfb2efca8714f5f2e0910befa6";
const std::string kExpectedPart1 = "394829";
const std::string kExpectedPart2 = "31150702";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day5/d5_input.txt", "Day5/d5_input.txt", "../Day5/d5_input.txt", "../../Day5/d5_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::vector<int> parse_offsets(const std::string& raw) {
    std::istringstream in(raw);
    std::vector<int> out;
    int v = 0;
    while (in >> v) out.push_back(v);
    return out;
}

long long run_program(const std::vector<int>& offsets, int part) {
    std::vector<int> a = offsets;
    long long steps = 0;
    int i = 0;
    while (i >= 0 && i < static_cast<int>(a.size())) {
        const int jump = a[static_cast<std::size_t>(i)];
        if (part == 1) {
            a[static_cast<std::size_t>(i)] = jump + 1;
        } else if (jump >= 3) {
            a[static_cast<std::size_t>(i)] = jump - 1;
        } else {
            a[static_cast<std::size_t>(i)] = jump + 1;
        }
        i += jump;
        steps++;
    }
    return steps;
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");
    const auto offsets = parse_offsets(raw);

    auto t0 = std::chrono::steady_clock::now();
    const std::string ans = std::to_string(run_program(offsets, part));
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=5 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day05.rs
// Architecture Notes:
// - This file is heavily commented for long-term maintainability and reconstruction.
// - Pipeline shape is parse -> model -> compute -> emit.
// - Performance-sensitive sections document data-layout and concurrency tradeoffs.

use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "e9c74e01657b99ad1be3cedce52f75bb0e2ac9dfb2efca8714f5f2e0910befa6";
const EXPECTED_PART1: &str = "394829";
const EXPECTED_PART2: &str = "31150702";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = ["advent2017/Day5/d5_input.txt", "Day5/d5_input.txt", "../Day5/d5_input.txt", "../../Day5/d5_input.txt"];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn parse_offsets(raw: &str) -> Vec<i32> {
    raw.lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                None
            } else {
                Some(line.parse::<i32>().expect("int"))
            }
        })
        .collect::<Vec<i32>>()
}

fn run_program(offsets: &[i32], part: i32) -> i64 {
    let mut a = offsets.to_vec();
    let mut i: i32 = 0;
    let mut steps: i64 = 0;

    while i >= 0 && (i as usize) < a.len() {
        let idx = i as usize;
        let jump = a[idx];
        if part == 1 {
            a[idx] += 1;
        } else if jump >= 3 {
            a[idx] -= 1;
        } else {
            a[idx] += 1;
        }
        i += jump;
        steps += 1;
    }

    steps
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => {
                part = args[i + 1].parse().unwrap();
                i += 2;
            }
            "--input" => {
                input = args[i + 1].clone();
                i += 2;
            }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 {
        panic!("--part 1|2 required");
    }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let offsets = parse_offsets(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let ans = run_program(&offsets, part).to_string();
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected {
        panic!("answer mismatch");
    }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=5 part={} runtime_ms={:.3}", part, ms);
}
